LLama/Model/ModelManager.cs

```diff
@@ -0,0 +1,309 @@
+    }
+
+    private void GetModelsFromDirectories(params string[] dirs)
+    {
+        foreach (var dir in dirs)
+        {
+            var fullDirectoryPath = Path.GetFullPath(dir);
+
+            if (!Directory.Exists(fullDirectoryPath))
+            {
+                Trace.TraceError($"Model directory '{fullDirectoryPath}' does not exist");
+                continue;
+            }
+
+            if (_availableModels.ContainsKey(fullDirectoryPath))
+            {
+                Trace.TraceWarning($"Model directory '{fullDirectoryPath}' already probed");
+                continue;
+            }
+
+            // find models in current dir that are of expected type
+            List<ModelFileMetadata> directoryModelFiles = [];
+            foreach (var file in Directory.EnumerateFiles(fullDirectoryPath))
+            {
+                if (!ExpectedModelFileTypes.Contains(Path.GetExtension(file)))
+                {
+                    continue;
+                }
+
+                // expected model file
+                var fi = new FileInfo(file);
+                directoryModelFiles.Add(new ModelFileMetadata
+                {
+                    FileName = fi.Name,
+                    FilePath = fi.FullName,
+                    ModelType = ModelFileType.GGUF,
+                    SizeInBytes = fi.Length,
+                });
+            }
+
+            _availableModels.Add(fullDirectoryPath, directoryModelFiles);
+        }
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<ModelFileMetadata> ModelFileList
+        => _availableModels.SelectMany(x => x.Value);
+    /// <inheritdoc />
+    public IEnumerable<string> ModelDirectories
+        => _availableModels.Keys;
+
+    /// <inheritdoc />
+    public void AddDirectory(string directory)
+    {
+        GetModelsFromDirectories(directory);
+    }
+
+    /// <inheritdoc />
+    public bool RemoveDirectory(string directory)
+    {
+        return _availableModels.Remove(Path.GetFullPath(directory));
+    }
+
+    /// <inheritdoc />
+    public void RemoveAllDirectories()
+    {
+        _availableModels.Clear();
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<ModelFileMetadata> GetAvailableModelsFromDirectory(string directory)
+    {
+        var dirPath = Path.GetFullPath(directory);
+        return _availableModels.TryGetValue(dirPath, out var dirModels)
+            ? dirModels
+            : [];
+    }
+
+    /// <inheritdoc />
+    public bool TryGetModelFileMetadata(string fileName, out ModelFileMetadata modelMeta)
+    {
+        var filePath = Path.GetFullPath(fileName);
+        modelMeta = ModelFileList.FirstOrDefault(f => f.FilePath == filePath)!;
+        return modelMeta != null;
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<LLamaWeights> GetLoadedModels()
+    {
+        return _loadedModelCache.Values;
+    }
+
+    /// <inheritdoc />
+    public bool TryGetLoadedModel(string modelId, out LLamaWeights model)
```

Comment:
One potential failure case I can see here is if someone were to retrieve weights from the `ModelManager` and then Dispose them. The cache would then contain disposed weights (which is fine) but then would return a disposed `LLamaWeights` object next time (which is not fine).

I think you probably need to add a check when you retrieve something from the `_loadedModelCache`, if the object has been disposed remove it from the cache and proceed from there as if it was never in the cache at all.

Reply:
good catch, addressed